src/Bicep.Core/Emit/TemplateEmitter.cs

```diff
@@ -24,44 +25,55 @@ public TemplateEmitter(SemanticModel model)
+            using (var streamWriter = new StreamWriter(memoryStream, UTF8EncodingWithoutBom))
+            using (var writer = new JsonTextWriter(streamWriter))
+            {
+                new TemplateWriter(writer, this.model).Write(true, assemblyVersion, null);
+            }
+            return TemplateHashExtensions.ComputeTemplateHash(UTF8EncodingWithoutBom.GetString(memoryStream.ToArray()));
+        }
+
         /// <summary>
         /// Emits a template to the specified stream if there are no errors. No writes are made to the stream if there are compilation errors.
         /// </summary>
         /// <param name="stream">The stream to write the template</param>
-        public EmitResult Emit(Stream stream) => EmitOrFail(() =>
+        public EmitResult Emit(Stream stream, string assemblyVersion = "dev") => EmitOrFail((string templateHash) =>
         {
             using var writer = new JsonTextWriter(new StreamWriter(stream, UTF8EncodingWithoutBom, 4096, true))
             {
                 Formatting = Formatting.Indented
             };
 
-            new TemplateWriter(writer, this.model).Write();
-        });
+            new TemplateWriter(writer, this.model).Write(true, assemblyVersion, templateHash);
+        }, assemblyVersion);
 
         /// <summary>
         /// Emits a template to the specified text writer if there are no errors. No writes are made to the writer if there are compilation errors.
         /// </summary>
         /// <param name="textWriter">The text writer to write the template</param>
-        public EmitResult Emit(TextWriter textWriter) => EmitOrFail(() =>
+        public EmitResult Emit(TextWriter textWriter, string assemblyVersion = "dev") => EmitOrFail((string templateHash) =>
         {
             using var writer = new JsonTextWriter(textWriter)
             {
                 Formatting = Formatting.Indented
             };
 
-            new TemplateWriter(writer, this.model).Write();
-        });
+            new TemplateWriter(writer, this.model).Write(true, assemblyVersion, templateHash);
+        }, assemblyVersion);
 
         /// <summary>
         /// Emits a template to the specified json writer if there are no errors. No writes are made to the writer if there are compilation errors.
         /// </summary>
         /// <param name="writer">The json writer to write the template</param>
-        public EmitResult Emit(JsonTextWriter writer) => this.EmitOrFail(() =>
+        public EmitResult Emit(JsonTextWriter writer, string assemblyVersion = "dev") => this.EmitOrFail((string templateHash) =>
         {
-            new TemplateWriter(writer, this.model).Write();
-        });
+            new TemplateWriter(writer, this.model).Write(true, assemblyVersion, templateHash);
+        }, assemblyVersion);
 
-        private EmitResult EmitOrFail(Action write)
+        private EmitResult EmitOrFail(Action<string> write, string assemblyVersion)
```

Comment:
Mind renaming `write` => `writeTemplateFunc`? Took me a while to notice this was a param and not a method on the class.